compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3214,12 +3214,8 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                     _ => return,
                 }
 
-                // Pass and return structures up to 2 pointers in size by value, matching `ScalarPair`.
-                // LLVM will usually pass these in 2 registers, which is more efficient than by-ref.
-                let max_by_val_size = Pointer.size(self) * 2;
                 let size = arg.layout.size;
-
-                if arg.layout.is_unsized() || size > max_by_val_size {
+                if arg.layout.is_unsized() || size > Pointer.size(self) {
```

Comment:
Hi. I've recompiled with this merge amended as requested:

```diff
@@ -3215,7 +3215,7 @@ fn fn_abi_adjust_for_abi(
                 }
 
                 let size = arg.layout.size;
-                if arg.layout.is_unsized() || size > Pointer.size(self) {
+                if arg.layout.is_unsized() || size > Pointer.size(self) * 2 {
                     arg.make_indirect();
                 } else {
                     // We want to pass small aggregates as immediates, but using
```
Unfortunately, this again breaks auto-vectorization, as can be seen below.

@dotdash makes an interesting comment [here](https://github.com/rust-lang/rust/issues/85265#issuecomment-989821465) regarding the underlying mechanisms.

Code:
```Rust
pub fn case_4_f32_add(a: [f32; 4], b: [f32; 4]) -> [f32; 4] {
    [a[0] + b[0], a[1] + b[1], a[2] + b[2], a[3] + b[3]]
}
```
This merge, as-is:
```asm
 autovec::case_4_i32_add:
 mov     rax, rdi
 movdqu  xmm0, xmmword, ptr, [rsi]
 movdqu  xmm1, xmmword, ptr, [rdx]
 paddd   xmm1, xmm0
 movdqu  xmmword, ptr, [rdi], xmm1
 ret
```
This merge, amended as above:
```asm
autovec::case_4_i32_add:
 mov     r8, rcx
 add     ecx, esi
 shr     rsi, 32
 lea     eax, [rdx, +, rdi]
 shr     rdi, 32
 shr     r8, 32
 shr     rdx, 32
 add     edx, edi
 add     esi, r8d
 shl     rsi, 32
 or      rcx, rsi
 shl     rdx, 32
 or      rax, rdx
 mov     rdx, rcx
 ret
```

Reply:
> Unfortunately, this again breaks auto-vectorization

Consider making a codegen test for this, since it sounds like we don't have one given the regression.

Something like this, maybe?
```rust
// compile-flags: -C opt-level=3 -C target-cpu=x86-64-v3
// only-x86_64

// CHECK-LABEL: @case_4_f32_add
#[no_mangle]
pub fn case_4_f32_add(a: [f32; 4], b: [f32; 4]) -> [f32; 4] {
    // CHECK: fadd <4 x float>
    [a[0] + b[0], a[1] + b[1], a[2] + b[2], a[3] + b[3]]
}
```

Reply:
>> Can you also try 2*pointer size just for return values like before #79547?
>
> Hi. I've recompiled with this merge amended as requested:
> 
> ```diff
> @@ -3215,7 +3215,7 @@ fn fn_abi_adjust_for_abi(
>                  }
>  
>                  let size = arg.layout.size;
> -                if arg.layout.is_unsized() || size > Pointer.size(self) {
> +                if arg.layout.is_unsized() || size > Pointer.size(self) * 2 {
>                      arg.make_indirect();
>                  } else {
>                      // We want to pass small aggregates as immediates, but using
> ```

I think you went too far -- that's essentially nullifying this entire PR, the same as `max_by_val_size` was for all arguments and return values. But the line *before* #79547 doubled it only for return values like this:

```rust
let max_by_val_size = if is_ret { Pointer.size(cx) * 2 } else { Pointer.size(cx) };
```

(with a little more code to pass `is_ret` here.)

> > Unfortunately, this again breaks auto-vectorization
> 
> Consider making a codegen test for this, since it sounds like we don't have one given the regression.

Yes, please.

Reply:
@cuviper @bjorn3 . Sorry for barking up the wrong tree. I've reworked and recompiled:

```diff
@@ -3176,7 +3176,7 @@ fn fn_abi_adjust_for_abi(
             || abi == SpecAbi::RustIntrinsic
             || abi == SpecAbi::PlatformIntrinsic
         {
-            let fixup = |arg: &mut ArgAbi<'tcx, Ty<'tcx>>| {
+            let fixup = |arg: &mut ArgAbi<'tcx, Ty<'tcx>>, is_ret: bool| {
                 if arg.is_ignore() {
                     return;
                 }
@@ -3214,8 +3214,13 @@ fn fn_abi_adjust_for_abi(
                     _ => return,
                 }
 
+                // Return structures up to 2 pointers in size by value, matching `ScalarPair`. LLVM
+                // will usually return these in 2 registers, which is more efficient than by-ref.
+                let max_by_val_size =
+                    if is_ret { Pointer.size(self) * 2 } else { Pointer.size(self) };
                 let size = arg.layout.size;
-                if arg.layout.is_unsized() || size > Pointer.size(self) {
+
+                if arg.layout.is_unsized() || size > max_by_val_size {
                     arg.make_indirect();
                 } else {
                     // We want to pass small aggregates as immediates, but using
@@ -3224,9 +3229,9 @@ fn fn_abi_adjust_for_abi(
                     arg.cast_to(Reg { kind: RegKind::Integer, size });
                 }
             };
-            fixup(&mut fn_abi.ret);
+            fixup(&mut fn_abi.ret, true);
             for arg in &mut fn_abi.args {
-                fixup(arg);
+                fixup(arg, false);
             }
         } else {
             fn_abi.adjust_for_foreign_abi(self, abi)?;
```

Sadly, the output is not much better than Rust 1.57.0.

Code:
```rust
pub fn case_4_f32_add(a: [f32; 4], b: [f32; 4]) -> [f32; 4] {
    [a[0] + b[0], a[1] + b[1], a[2] + b[2], a[3] + b[3]]
}
```

This merge, as-is:
```asm
 autovec::case_4_f32_add:
 mov     rax, rdi
 movups  xmm0, xmmword, ptr, [rsi]
 movups  xmm1, xmmword, ptr, [rdx]
 addps   xmm1, xmm0
 movups  xmmword, ptr, [rdi], xmm1
 ret
```

This merge, amended as above:
```asm
 autovec::case_4_f32_add:
 movss   xmm0, dword, ptr, [rdi]
 addss   xmm0, dword, ptr, [rsi]
 movss   xmm1, dword, ptr, [rdi, +, 8]
 addss   xmm1, dword, ptr, [rsi, +, 8]
 movups  xmm2, xmmword, ptr, [rdi]
 movups  xmm3, xmmword, ptr, [rsi]
 movss   xmm4, dword, ptr, [rdi, +, 4]
 shufps  xmm4, xmm2, 244
 shufps  xmm4, xmm2, 242
 movss   xmm2, dword, ptr, [rsi, +, 4]
 shufps  xmm2, xmm3, 244
 shufps  xmm2, xmm3, 242
 addps   xmm2, xmm4
 movd    eax, xmm0
 movd    ecx, xmm1
 movd    edx, xmm2
 shufps  xmm2, xmm2, 85
 movd    esi, xmm2
 shl     rsi, 32
 shl     rdx, 32
 or      rdx, rcx
 or      rax, rsi
 ret
```

Rust 1.57.0
```asm
 autovec::case_4_f32_add:
 movd    xmm0, esi
 mov     rax, rsi
 shld    rax, rdi, 32
 shr     rsi, 32
 movq    xmm1, rax
 movq    xmm2, rsi
 punpckldq xmm2, xmm1
 movd    xmm1, edi
 movd    xmm3, edx
 addps   xmm3, xmm1
 movd    xmm1, ecx
 addss   xmm1, xmm0
 shrd    rdx, rcx, 32
 shr     rcx, 32
 movq    xmm0, rdx
 movq    xmm4, rcx
 punpckldq xmm4, xmm0
 addps   xmm4, xmm2
 movd    eax, xmm3
 movd    ecx, xmm1
 movd    edx, xmm4
 shufps  xmm4, xmm4, 85
 movd    esi, xmm4
 shl     rsi, 32
 shl     rdx, 32
 or      rdx, rcx
 or      rax, rsi
 ret
```

Reply:
It'd be great if you could also include the LLVM IR with the assembly code that makes it easier to see where the problem is. Looking at the assembly code, I guess the return type is still i128. That means that it uses integer registers which means that it has to extract the values out of the SSE registers. Inputs seem to be through memory again, but the need for the results to go into integer registers seems to cause LLVM to perform the additions in an odd way.

To get proper results here the Rust ABI code needs to get a bit more sophisticated. The C ABI handlers are a bit better but from what I could gather during the weekend, its underlying machinery also can't handle things the way LLVM needs them to produce the best possible code, like passing a struct or array that has four floats as two vectors of two float each (for x86_64 SysV).

NB: I didn't check whether it already does that, but the code that handles the automatic  memory layout for aggregates might also need to take ABI details into consideration, e.g. to put floats next to each other to make them eligible for passing in a single SSE register etc.

Reply:
I think that it's liable we will get the opportunity to improve here in the future, and we should open a new issue collecting some of the accreted data around this matter if we want to make further improvements. Recovering the vectorization in this _relatively_ easy, "my first compiler's loop autovectorization" case and adding a regression test to confirm it will likely accelerate future attempts more than hammering at this particular spot in the code.

Reply:
@dotdash 
> It'd be great if you could also include the LLVM IR with the assembly code that makes it easier to see where the problem is.

The LLVM IR/ x64 assembly outputs for the [reworked](https://github.com/rust-lang/rust/pull/91719#discussion_r769088827) amendments above.

Code:
```rust
pub fn case_4_f32_add(a: [f32; 4], b: [f32; 4]) -> [f32; 4] {
    [a[0] + b[0], a[1] + b[1], a[2] + b[2], a[3] + b[3]]
}
```

This merge, as-is:
```llvm
; ModuleID = 'autovec.47b73f22-cgu.0'
source_filename = "autovec.47b73f22-cgu.0"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

; autovec::case_4_f32_add
; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind uwtable willreturn
define void @_ZN7autovec14case_4_f32_add17hdb54e55ec6e798beE([4 x float]* noalias nocapture sret([4 x float]) dereferenceable(16) %0, [4 x float]* noalias nocapture readonly dereferenceable(16) %a, [4 x float]* noalias nocapture readonly dereferenceable(16) %b) unnamed_addr #0 {
start:
  %1 = bitcast [4 x float]* %a to <4 x float>*
  %2 = load <4 x float>, <4 x float>* %1, align 4
  %3 = bitcast [4 x float]* %b to <4 x float>*
  %4 = load <4 x float>, <4 x float>* %3, align 4
  %5 = fadd <4 x float> %2, %4
  %6 = bitcast [4 x float]* %0 to <4 x float>*
  store <4 x float> %5, <4 x float>* %6, align 4
  ret void
}

attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind uwtable willreturn "probe-stack"="__rust_probestack" "target-cpu"="skylake" }

!llvm.module.flags = !{!0, !1}

!0 = !{i32 7, !"PIC Level", i32 2}
!1 = !{i32 2, !"RtLibUseGOT", i32 1}
```
```asm
 autovec::case_4_f32_add:
 mov     rax, rdi
 movups  xmm0, xmmword, ptr, [rsi]
 movups  xmm1, xmmword, ptr, [rdx]
 addps   xmm1, xmm0
 movups  xmmword, ptr, [rdi], xmm1
 ret
```
This merge, amended as above:
```llvm
; ModuleID = 'autovec.47b73f22-cgu.0'
source_filename = "autovec.47b73f22-cgu.0"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

; autovec::case_4_f32_add
; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readonly uwtable willreturn
define i128 @_ZN7autovec14case_4_f32_add17hdb54e55ec6e798beE([4 x float]* noalias nocapture readonly dereferenceable(16) %a, [4 x float]* noalias nocapture readonly dereferenceable(16) %b) unnamed_addr #0 {
start:
  %0 = getelementptr inbounds [4 x float], [4 x float]* %a, i64 0, i64 0
  %_4 = load float, float* %0, align 4
  %1 = getelementptr inbounds [4 x float], [4 x float]* %b, i64 0, i64 0
  %_6 = load float, float* %1, align 4
  %_3 = fadd float %_4, %_6
  %2 = getelementptr inbounds [4 x float], [4 x float]* %a, i64 0, i64 1
  %3 = getelementptr inbounds [4 x float], [4 x float]* %b, i64 0, i64 1
  %4 = bitcast float* %2 to <2 x float>*
  %5 = load <2 x float>, <2 x float>* %4, align 4
  %6 = bitcast float* %3 to <2 x float>*
  %7 = load <2 x float>, <2 x float>* %6, align 4
  %8 = fadd <2 x float> %5, %7
  %9 = getelementptr inbounds [4 x float], [4 x float]* %a, i64 0, i64 3
  %_19 = load float, float* %9, align 4
  %10 = getelementptr inbounds [4 x float], [4 x float]* %b, i64 0, i64 3
  %_21 = load float, float* %10, align 4
  %_18 = fadd float %_19, %_21
  %11 = bitcast float %_3 to i32
  %12 = bitcast <2 x float> %8 to <2 x i32>
  %13 = bitcast float %_18 to i32
  %.sroa.4.0.insert.ext = zext i32 %13 to i128
  %.sroa.4.0.insert.shift = shl nuw i128 %.sroa.4.0.insert.ext, 96
  %14 = zext <2 x i32> %12 to <2 x i128>
  %15 = shl nuw nsw <2 x i128> %14, <i128 32, i128 64>
  %16 = extractelement <2 x i128> %15, i32 1
  %.sroa.3.0.insert.insert = or i128 %.sroa.4.0.insert.shift, %16
  %17 = extractelement <2 x i128> %15, i32 0
  %.sroa.2.0.insert.insert = or i128 %.sroa.3.0.insert.insert, %17
  %.sroa.0.0.insert.ext = zext i32 %11 to i128
  %.sroa.0.0.insert.insert = or i128 %.sroa.2.0.insert.insert, %.sroa.0.0.insert.ext
  ret i128 %.sroa.0.0.insert.insert
}

attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readonly uwtable willreturn "probe-stack"="__rust_probestack" "target-cpu"="skylake" }

!llvm.module.flags = !{!0, !1}

!0 = !{i32 7, !"PIC Level", i32 2}
!1 = !{i32 2, !"RtLibUseGOT", i32 1}
```
```asm
 autovec::case_4_f32_add:
 movss   xmm0, dword, ptr, [rdi]
 addss   xmm0, dword, ptr, [rsi]
 movss   xmm1, dword, ptr, [rdi, +, 8]
 addss   xmm1, dword, ptr, [rsi, +, 8]
 movups  xmm2, xmmword, ptr, [rdi]
 movups  xmm3, xmmword, ptr, [rsi]
 movss   xmm4, dword, ptr, [rdi, +, 4]
 shufps  xmm4, xmm2, 244
 shufps  xmm4, xmm2, 242
 movss   xmm2, dword, ptr, [rsi, +, 4]
 shufps  xmm2, xmm3, 244
 shufps  xmm2, xmm3, 242
 addps   xmm2, xmm4
 movd    eax, xmm0
 movd    ecx, xmm1
 movd    edx, xmm2
 shufps  xmm2, xmm2, 85
 movd    esi, xmm2
 shl     rsi, 32
 shl     rdx, 32
 or      rdx, rcx
 or      rax, rsi
 ret
```
Rust 1.57.0
```llvm
; ModuleID = 'autovec.990a041e-cgu.0'
source_filename = "autovec.990a041e-cgu.0"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

; autovec::case_4_f32_add
; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn
define i128 @_ZN7autovec14case_4_f32_add17h1bb821e346d15ffaE(i128 %0, i128 %1) unnamed_addr #0 {
start:
  %.sroa.5.0.extract.shift = lshr i128 %0, 64
  %.sroa.5.0.extract.trunc = trunc i128 %.sroa.5.0.extract.shift to i32
  %2 = bitcast i32 %.sroa.5.0.extract.trunc to float
  %3 = insertelement <2 x i128> poison, i128 %0, i32 0
  %4 = shufflevector <2 x i128> %3, <2 x i128> poison, <2 x i32> zeroinitializer
  %5 = lshr <2 x i128> %4, <i128 96, i128 32>
  %6 = trunc <2 x i128> %5 to <2 x i32>
  %7 = bitcast <2 x i32> %6 to <2 x float>
  %8 = insertelement <2 x i128> %3, i128 %1, i32 1
  %9 = trunc <2 x i128> %8 to <2 x i32>
  %10 = bitcast <2 x i32> %9 to <2 x float>
  %.sroa.59.0.extract.shift = lshr i128 %1, 64
  %.sroa.59.0.extract.trunc = trunc i128 %.sroa.59.0.extract.shift to i32
  %11 = bitcast i32 %.sroa.59.0.extract.trunc to float
  %12 = insertelement <2 x i128> poison, i128 %1, i32 0
  %13 = shufflevector <2 x i128> %12, <2 x i128> poison, <2 x i32> zeroinitializer
  %14 = lshr <2 x i128> %13, <i128 96, i128 32>
  %15 = trunc <2 x i128> %14 to <2 x i32>
  %16 = bitcast <2 x i32> %15 to <2 x float>
  %shift = shufflevector <2 x float> %10, <2 x float> poison, <2 x i32> <i32 1, i32 undef>
  %17 = fadd <2 x float> %shift, %10
  %_13 = fadd float %2, %11
  %18 = fadd <2 x float> %7, %16
  %bc = bitcast <2 x float> %17 to <2 x i32>
  %19 = extractelement <2 x i32> %bc, i32 0
  %20 = bitcast float %_13 to i32
  %21 = bitcast <2 x float> %18 to <2 x i32>
  %.sroa.3.0.insert.ext = zext i32 %20 to i128
  %.sroa.3.0.insert.shift = shl nuw nsw i128 %.sroa.3.0.insert.ext, 64
  %22 = zext <2 x i32> %21 to <2 x i128>
  %23 = shl nuw <2 x i128> %22, <i128 96, i128 32>
  %24 = extractelement <2 x i128> %23, i32 0
  %.sroa.3.0.insert.insert = or i128 %24, %.sroa.3.0.insert.shift
  %25 = extractelement <2 x i128> %23, i32 1
  %.sroa.2.0.insert.insert = or i128 %.sroa.3.0.insert.insert, %25
  %.sroa.013.0.insert.ext = zext i32 %19 to i128
  %.sroa.013.0.insert.insert = or i128 %.sroa.2.0.insert.insert, %.sroa.013.0.insert.ext
  ret i128 %.sroa.013.0.insert.insert
}

attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn "probe-stack"="__rust_probestack" "target-cpu"="skylake" }

!llvm.module.flags = !{!0, !1}

!0 = !{i32 7, !"PIC Level", i32 2}
!1 = !{i32 2, !"RtLibUseGOT", i32 1}
```
```asm
 autovec::case_4_f32_add:
 movd    xmm0, esi
 mov     rax, rsi
 shld    rax, rdi, 32
 shr     rsi, 32
 movq    xmm1, rax
 movq    xmm2, rsi
 punpckldq xmm2, xmm1
 movd    xmm1, edi
 movd    xmm3, edx
 addps   xmm3, xmm1
 movd    xmm1, ecx
 addss   xmm1, xmm0
 shrd    rdx, rcx, 32
 shr     rcx, 32
 movq    xmm0, rdx
 movq    xmm4, rcx
 punpckldq xmm4, xmm0
 addps   xmm4, xmm2
 movd    eax, xmm3
 movd    ecx, xmm1
 movd    edx, xmm4
 shufps  xmm4, xmm4, 85
 movd    esi, xmm4
 shl     rsi, 32
 shl     rdx, 32
 or      rdx, rcx
 or      rax, rsi
 ret
```